pyteal/ast/scratch.py

```diff
@@ -9,7 +9,58 @@
     from ..compiler import CompileOptions
 
 
-class ScratchSlot:
+class Slot:
+    """Abstract Slot class"""
+
+    def slotIdFromStack(self) -> bool:
+        pass
+
+    def store(self, value: Expr = None) -> Expr:
+        """Get an expression to store a value in this slot.
+
+        Args:
+            value (optional): The value to store in this slot. If not included, the last value on
+            the stack will be stored. NOTE: storing the last value on the stack breaks the typical
+            semantics of PyTeal, only use if you know what you're doing.
+        """
+        if value is not None:
+            return ScratchStore(self, value)
+        return ScratchStackStore(self)
+
+    def load(self, type: TealType = TealType.anytype) -> "ScratchLoad":
+        """Get an expression to load a value from this slot.
+
+        Args:
+            type (optional): The type being loaded from this slot, if known. Defaults to
+                TealType.anytype.
+        """
+        return ScratchLoad(self, type)
+
+
+Slot.__module__ = "pyteal"
+
+
+class DynamicSlot(Slot):
```

Comment:
I like this direction, but still I don't see how `DynamicSlot` knows which slot it should store to/load from.

In my opinion, `DynamicSlot` needs a constructor that takes an `Expr` that indicates this value.

Reply:
Will keep this in mind

Reply:
I think this isn't needed. Imagine we are trying store the value `val: Expr`. Let's assume that `ScratchVar.slot: Union[DynamicSlot, ScratchSlot]` is set during `ScratchVar`'s construction based on `slotId`'s type and follow the code path:
```python
ScrachVar(slotId).store(val) -> dispatches ->
    (DynamicSlot|ScratchSlot).store(val) ->
        ScratchStore(val) : Expr
```
So in either case, we end up staring `val` via `ScratchStore(val)`